File: src/lib/tensorkv/timing.ts
/** Named-part timing: serialization, handle install, prefill vs decode compute. */

import {
  COMPUTE_MS_AT_32K,
  DEFAULT_CREDIT_GBPS,
  HANDLE_INSTALL_NS,
  PREFILL_COMPUTE_MS_AT_32K,
  PREFILL_TOKENS,
  TOKENS_PER_BLOCK,
} from "./types";

export const TKV_NETWORK_NS = 800;
export const TKV_RMT_NS = 150;
export const TKV_PCIE_DMA_NS = 1150;

export function serializeNs(nBytes: number, gbps: number) {
  if (nBytes <= 0 || gbps <= 0) return 0;
  return Math.trunc((nBytes * 8) / gbps);
}

export function serializeMs(nBytes: number, gbps: number) {
  return serializeNs(nBytes, gbps) / 1e6;
}

export function attentionComputeMs(nTokens: number) {
  return (COMPUTE_MS_AT_32K * Math.max(0, nTokens)) / PREFILL_TOKENS;
}

export function prefillComputeMs(nTokens: number) {
  return (PREFILL_COMPUTE_MS_AT_32K * Math.max(0, nTokens)) / PREFILL_TOKENS;
}

export function handleInstallMs(nTokens: number) {
  const nBlocks = Math.max(1, Math.ceil(Math.max(0, nTokens) / TOKENS_PER_BLOCK));
  return (nBlocks * HANDLE_INSTALL_NS) / 1e6;
}

export type TTFTParts = {
  setupMs: number;
  fetchMs: number;
  computeMs: number;
  totalMs: number;
};

export function ttftBreakdown(args: {
  prefixHit: boolean;
  prefixTokens: number;
  localTokens: number;
  gatheredBytes: number;
  getLatencyNs: number;
  probeLatencyNs: number;
  creditGbps?: number;
}): TTFTParts {
  const credit = args.creditGbps ?? DEFAULT_CREDIT_GBPS;
  const contextTokens = args.prefixHit
    ? args.prefixTokens + Math.max(0, args.localTokens)
    : Math.max(args.prefixTokens, args.localTokens);
  let setupMs = args.probeLatencyNs / 1e6;
  let computeMs: number;
  if (args.prefixHit) {
    setupMs += handleInstallMs(args.prefixTokens);
    computeMs = attentionComputeMs(Math.max(contextTokens, args.prefixTokens));
  } else {
    computeMs = prefillComputeMs(Math.max(args.localTokens, args.prefixTokens, 1));
  }
  const fetchMs = serializeMs(args.gatheredBytes, credit) + args.getLatencyNs / 1e6;
  return { setupMs, fetchMs, computeMs, totalMs: setupMs + fetchMs + computeMs };
}

File: python/tensorkv/timing.py
"""First-principles timing for the software replica.

FPGA wall-clock numbers from the paper stay in ``constants.PAPER_*``.
These helpers compose serialization delay, SRAM/HBM pipeline time, and
attention compute scaled from the paper's 32K-token 15 ms kernel.
"""

from __future__ import annotations

from dataclasses import dataclass

from .constants import (
    DEFAULT_CREDIT_GBPS,
    LINK_GBPS,
    PAPER_COMPUTE_MS_AT_32K,
    PAPER_PREFILL_TOKENS,
)


def serialize_ns(n_bytes: int, gbps: float) -> int:
    """Wire serialization time for `n_bytes` at `gbps` gigabits/sec."""
    if n_bytes <= 0 or gbps <= 0:
        return 0
    return int(n_bytes * 8 / gbps)


def serialize_ms(n_bytes: int, gbps: float) -> float:
    return serialize_ns(n_bytes, gbps) / 1e6


def attention_compute_ms(n_tokens: int) -> float:
    """Scale the paper's 15 ms / 32K-token kernel to `n_tokens`."""
    return PAPER_COMPUTE_MS_AT_32K * max(0, n_tokens) / PAPER_PREFILL_TOKENS


@dataclass
class TTFTParts:
    setup_ms: float
    fetch_ms: float
    compute_ms: float
    total_ms: float


def ttft_breakdown(
    *,
    prefix_hit: bool,
    prefix_tokens: int,
    local_tokens: int,
    gathered_bytes: int,
    get_latency_ns: int,
    probe_latency_ns: int,
    credit_gbps: float = DEFAULT_CREDIT_GBPS,
    link_gbps: float = LINK_GBPS,
) -> TTFTParts:
    """Compose TTFT from control RTT + credit-shaped gather + local compute.

    Prefix hit: PROBE returns handles only (no HBM); compute covers the suffix.
    Prefix miss: local prefill of the whole prompt, then gather.
    """
    descriptor = 64
    setup_ms = serialize_ms(descriptor, link_gbps) + probe_latency_ns / 1e6
    if prefix_hit:
        setup_ms += serialize_ms(max(1, prefix_tokens) * 8, link_gbps)
    compute_ms = attention_compute_ms(local_tokens)
    fetch_ms = serialize_ms(gathered_bytes, credit_gbps) + get_latency_ns / 1e6
    total = setup_ms + fetch_ms + compute_ms
    return TTFTParts(setup_ms, fetch_ms, compute_ms, total)
